// @uiw/codemirror-theme-solarized/cjs/light.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault")["default"];
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.solarizedLightInit = exports.solarizedLight = exports.defaultSettingsSolarizedLight = void 0;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _highlight = require("@lezer/highlight");
var _codemirrorThemes = require("@uiw/codemirror-themes");
var _lightColor = require("./light-color");
var defaultSettingsSolarizedLight = exports.defaultSettingsSolarizedLight = {
  background: _lightColor.config.background,
  foreground: _lightColor.config.foreground,
  caret: _lightColor.config.cursor,
  selection: _lightColor.config.selection,
  selectionMatch: _lightColor.config.selectionMatch,
  gutterBackground: _lightColor.config.background,
  gutterForeground: _lightColor.config.foreground,
  gutterBorder: 'transparent',
  lineHighlight: _lightColor.config.activeLine
};
var solarizedLightInit = exports.solarizedLightInit = function solarizedLightInit(options) {
  var _ref = options || {},
    _ref$theme = _ref.theme,
    theme = _ref$theme === void 0 ? 'light' : _ref$theme,
    _ref$settings = _ref.settings,
    settings = _ref$settings === void 0 ? {} : _ref$settings,
    _ref$styles = _ref.styles,
    styles = _ref$styles === void 0 ? [] : _ref$styles;
  return (0, _codemirrorThemes.createTheme)({
    theme: theme,
    settings: (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, defaultSettingsSolarizedLight), settings),
    styles: [{
      tag: _highlight.tags.keyword,
      color: _lightColor.config.keyword
    }, {
      tag: [_highlight.tags.name, _highlight.tags.deleted, _highlight.tags.character, _highlight.tags.macroName],
      color: _lightColor.config.variable
    }, {
      tag: [_highlight.tags.propertyName],
      color: _lightColor.config["function"]
    }, {
      tag: [_highlight.tags.processingInstruction, _highlight.tags.string, _highlight.tags.inserted, _highlight.tags.special(_highlight.tags.string)],
      color: _lightColor.config.string
    }, {
      tag: [_highlight.tags["function"](_highlight.tags.variableName), _highlight.tags.labelName],
      color: _lightColor.config["function"]
    }, {
      tag: [_highlight.tags.color, _highlight.tags.constant(_highlight.tags.name), _highlight.tags.standard(_highlight.tags.name)],
      color: _lightColor.config.constant
    }, {
      tag: [_highlight.tags.definition(_highlight.tags.name), _highlight.tags.separator],
      color: _lightColor.config.variable
    }, {
      tag: [_highlight.tags.className],
      color: _lightColor.config["class"]
    }, {
      tag: [_highlight.tags.number, _highlight.tags.changed, _highlight.tags.annotation, _highlight.tags.modifier, _highlight.tags.self, _highlight.tags.namespace],
      color: _lightColor.config.number
    }, {
      tag: [_highlight.tags.typeName],
      color: _lightColor.config.type,
      fontStyle: _lightColor.config.type
    }, {
      tag: [_highlight.tags.operator, _highlight.tags.operatorKeyword],
      color: _lightColor.config.keyword
    }, {
      tag: [_highlight.tags.url, _highlight.tags.escape, _highlight.tags.regexp, _highlight.tags.link],
      color: _lightColor.config.regexp
    }, {
      tag: [_highlight.tags.meta, _highlight.tags.comment],
      color: _lightColor.config.comment
    }, {
      tag: _highlight.tags.tagName,
      color: _lightColor.config.tag
    }, {
      tag: _highlight.tags.strong,
      fontWeight: 'bold'
    }, {
      tag: _highlight.tags.emphasis,
      fontStyle: 'italic'
    }, {
      tag: _highlight.tags.link,
      textDecoration: 'underline'
    }, {
      tag: _highlight.tags.heading,
      fontWeight: 'bold',
      color: _lightColor.config.heading
    }, {
      tag: [_highlight.tags.atom, _highlight.tags.bool, _highlight.tags.special(_highlight.tags.variableName)],
      color: _lightColor.config.variable
    }, {
      tag: _highlight.tags.invalid,
      color: _lightColor.config.invalid
    }, {
      tag: _highlight.tags.strikethrough,
      textDecoration: 'line-through'
    }].concat((0, _toConsumableArray2["default"])(styles))
  });
};
var solarizedLight = exports.solarizedLight = solarizedLightInit();
